lib/intf.js

```diff
@@ -267,26 +290,65 @@ export default function (self, ctor) {
     }
     return f;
   };
-  self.applyFilter = function () {
-    self.refreshFromOrigialData();
-    Object.keys(self.columnFilters).forEach(function (filter) {
-      var header = self.getHeaderByName(filter);
+  self.getFilteredAndSortedViewData = function (originalData) {
+    // We make a copy of originalData here in order be able to
+    // filter and sort rows without modifying the original array.
+    // Each row is turned into a (row, rowIndex) tuple
+    // so that when we apply filters, we can refer back to the
+    // row's original row number in originalData. This becomes
+    // useful when emitting cell events.
+    let newViewData = originalData.map((row, originalRowIndex) => [
+      row,
+      originalRowIndex,
+    ]);
+
+    // Apply filtering
+    for (const [headerName, filterText] of Object.entries(self.columnFilters)) {
+      const header = self.getHeaderByName(headerName);
+
       if (!header) {
-        return;
+        continue;
       }
-      self.currentFilter =
+
+      const currentFilterFunction =
         header.filter || self.filter(header.type || 'string');
-      self.data = self.data.filter(function (row) {
-        return self.currentFilter(row[filter], self.columnFilters[filter]);
+
+      newViewData = newViewData.filter(function ([row]) {
+        const cellValue = row[headerName];
+        const shouldIncludeRow = currentFilterFunction(cellValue, filterText);
+
+        return shouldIncludeRow;
       });
-    });
+    }
+
+    // Apply sorting
+    for (const column of self.orderings.columns) {
+      const sortFn = column.sortFunction(column.orderBy, column.orderDirection);
+
+      newViewData.sort(([rowA], [rowB]) => sortFn(rowA, rowB));
+    }
+
+    return {
+      viewData: newViewData.map(([row]) => row),
+      boundRowIndexMap: new Map(
+        newViewData.map(([_row, originalRowIndex], viewRowIndex) => [
+          viewRowIndex,
+          originalRowIndex,
+        ]),
+      ),
+    };
+  };
+  self.refresh = function () {
+    const { viewData, boundRowIndexMap } = self.getFilteredAndSortedViewData(
+      self.originalData,
+    );
+
+    self.viewData = viewData;
```

Comment:
`viewData` or.... `view`?  Since `view` isn't taken and `viewData` is a mouthful what do you think about changing this to `view`?

Reply:
I tend to err on the side overly verbose variable names, in case I forget three weeks down the line what it was I was doing. To me `view` is too vague in that it doesn't quite describe the nature of the variable. I agree that `viewData` is a mouthful, but feel it's easier to understand `grid.viewData[0]` as referring to the first row than `grid.view[0]` is.